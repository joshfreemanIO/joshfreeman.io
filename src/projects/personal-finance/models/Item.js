import { types, getSnapshot, applySnapshot } from 'mobx-state-tree'

const Item = types
  .model({
    name: '',
    value: 0,
    frequency: '',
    category: '',
    taxableIncomeReducing: false
  })
  .actions(self => ({
    set(name, value) {
      console.log(name, value)
      self[name] = value
    },
    updateValue(event) {
      const value = Number.parseInt(event.target.value)

      if (Number.isNaN(value)) {
        self.value = 0
      } else {
        self.value = value
      }
    }
  }))
  .views(self => ({
  }))

const Editable = types
  .model()
  .volatile(self => ({
    isEditing: false,
    previousSnapshot: {}
  }))
  .actions(self => ({
    cancel() {
      applySnapshot(self, self.previousSnapshot)
      self.isEditing = false
    },
    save() {
      self.previousSnapshot = {}
      self.isEditing = false
    },
    edit() {
      self.previousSnapshot = getSnapshot(self)
      self.isEditing = true
    },
    update(event) {
      self.set(event.target.name, event.target.value)
    }
  }))

export default types.compose(Item, Editable).named('Item')
