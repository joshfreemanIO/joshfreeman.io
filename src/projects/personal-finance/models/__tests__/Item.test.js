import Item from '../Item'
import { getSnapshot } from 'mobx-state-tree'

test('Item instantiation', () => {
  const item = Item.create()

  expect(getSnapshot(item)).toMatchSnapshot()
})
