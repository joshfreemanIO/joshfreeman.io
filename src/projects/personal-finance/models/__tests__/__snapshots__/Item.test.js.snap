// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Item instantiation 1`] = `
Object {
  "category": "",
  "frequency": "",
  "name": "",
  "taxableIncomeReducing": false,
  "value": 0,
}
`;
